### PR TITLE
Update facilitator course management to grant facilitator permission

### PR DESCRIPTION
**Pull request by sureshc**
Previously the Facilitator Course management functionality only searched for and assigned/removed courses from users who already had the facilitator permission.  Update this tool to allow search for any type of user and to dynamically assign facilitator permission when assigning the first course to a user.

workshop facilitator course assignment

@mehalshah @clareconstantine Can you review these changes?

### CODE REVIEW

File: dashboard/app/controllers/pd/workshop_user_management_controller.rb [hunk 1]
```diff
@@ -1,35 +1,68 @@
...
   end
 
   # post /pd/workshop_user_management/assign_course
   def assign_course
-    User.find(params[:facilitator_id]).try(:course_as_facilitator=, params[:course])
-    redirect_to action: "facilitator_courses_form", search_term: params[:facilitator_id]
+    @user = restricted_users.find_by(id: params[:user_id])
+    if @user.try(:teacher?)
+      @user.course_as_facilitator = params[:course]
+      @user.permission = UserPermission::FACILITATOR unless @user.facilitator?
```
Comment (clareconstantine) [inline dashboard/app/controllers/pd/workshop_user_management_controller.rb]: Line 50 does the same thing, but without `unless @user.facilitator?`. I'm not sure whether the check is needed, but it seems like these should be consistent.

File: dashboard/app/controllers/pd/workshop_user_management_controller.rb [hunk 2]
```diff
@@ -1,35 +1,68 @@
...
-        hashed_email: User.hash_email(search_term),
-        user_permissions: {permission: UserPermission::FACILITATOR}
-      )
+      @user = restricted_users.find_by(hashed_email: restricted_users.hash_email(search_term))
     end
 
-    unless @facilitator || search_term.blank?
-      flash[:notice] = "Facilitator not found"
+    unless @user || search_term.blank?
+      flash[:notice] = "User not found"
```
Comment (mehalshah) [inline dashboard/app/controllers/pd/workshop_user_management_controller.rb]: Do a notice with something like `No user with email <email> found`

File: dashboard/app/controllers/pd/workshop_user_management_controller.rb [hunk 3]
```diff
@@ -1,35 +1,68 @@
...
     end
   end
 
   # post /pd/workshop_user_management/assign_course
   def assign_course
```
Comment (mehalshah) [inline dashboard/app/controllers/pd/workshop_user_management_controller.rb]: Be more verbose in the method name - you are assigning a course for a facilitator and the name should reflect that

File: dashboard/app/controllers/pd/workshop_user_management_controller.rb [hunk 4]
```diff
@@ -1,35 +1,68 @@
...
+    redirect_to action: "facilitator_courses_form", search_term: params[:user_id]
   end
 
   # get /pd/workshop_user_management/remove_course
   def remove_course
-    User.find(params[:facilitator_id]).try(:delete_course_as_facilitator, params[:course])
-    redirect_to action: "facilitator_courses_form", search_term: params[:facilitator_id]
+    restricted_users.find(params[:user_id]).try(:delete_course_as_facilitator, params[:course])
+    redirect_to action: "facilitator_courses_form", search_term: params[:user_id]
+  end
```
Comment (mehalshah) [inline dashboard/app/controllers/pd/workshop_user_management_controller.rb]: Changed my mind - we should remove the Facilitator user permission if there are no more courses here

File: dashboard/app/controllers/pd/workshop_user_management_controller.rb [hunk 5]
```diff
@@ -1,35 +1,68 @@
...
+
+  # get /pd/workshop_user_management/update_facilitator_permission
+  def update_facilitator_permission
+    @user = restricted_users.find_by(id: params[:user_id])
+    if @user.try(:teacher?)
```
Comment (mehalshah) [inline dashboard/app/controllers/pd/workshop_user_management_controller.rb]: Why `try` here? If @user is nil, find_by will have already thrown an exception.

I think this method is unnecessary - we shouldn't be designating anyone as a facilitator without assigning a certain course to them. And if we want to de-facilitator-them, then remove all the course facilitators (or have a button in the UI that destroys all of them).

**Comment by sureshc** [comment]
@clareconstantine @mehalshah I've implemented your suggested revisions.  Can you take another look?

**Comment by sureshc** [comment]
Tests passed in #17760.

**Comment by sureshc** [comment]
Tests passed in #17760.

**Pull request merged**